### PR TITLE
Dev GroupNorm

### PR DESCRIPTION
**Pull request by MARD1NO**
```python
def GroupNorm(x, gamma, beta): 
    N, C, H, W = x.shape
    x = np.reshape(x, (N, G, C//G, H, W)
    mean = np.mean(x, (2, 3, 4))
    var = np.var(x, (2, 3, 4))
    normalized = (x - mean) / np.sqrt(var + eps)
    result = normalized * gamma + beta
```

### CODE REVIEW

File: oneflow/core/functional/impl/nn_functor.cpp [hunk 1]
```diff
@@ -781,6 +781,44 @@ class LayerNormAffineFunctor {
...
+class GroupNormFunctor {
+ public:
+  GroupNormFunctor() {
+    op_ = CHECK_JUST(one::OpBuilder("group_norm")
+                         .Input("x")
+                         .Output("y")
+                         .Output("mean")
+                         .Output("inv_variance")
+                         .Build());
+    affine_op_ = CHECK_JUST(one::OpBuilder("group_norm")
```
Comment (MARD1NO) [inline oneflow/core/functional/impl/nn_functor.cpp:793]: 遵循pytorch那边，affine就是同时存在gamma beta

File: oneflow/user/kernels/group_norm_kernel.cu [hunk 2]
```diff
@@ -0,0 +1,611 @@
...
+  int64_t channel_size;
+  int64_t spatial_size;
+};
+
+template<typename LOAD, typename STORE, typename ComputeType>
+struct DispatchGroupNormWarpImplPackSize {
+  cudaError_t operator()(cudaStream_t stream, LOAD load, STORE store, const int64_t rows,
+                         const int64_t cols, const int64_t spatial_size, const double epsilon,
+                         ComputeType* mean, ComputeType* inv_variance) {
+    if (spatial_size % 4 == 0) {
```
Comment (MARD1NO) [inline oneflow/user/kernels/group_norm_kernel.cu:110]: 这里Packsize取决于SpatialSize即 H x W。

输入是 N C H W，而gamma是 1 C 1 1。对连续的元素缩放的时候，我们希望这一串连续元素对应的gamma是同一个，因此需要通过spatialsize来限制其packsize

File: python/oneflow/test/modules/test_global_cumprod.py [hunk 3]
```diff
@@ -27,7 +27,7 @@ def _test_cumprod_impl(test_case, ndim, placement, sbp):
     x = random_tensor(ndim, *dims)
     y = x.to_global(placement=placement, sbp=sbp)
     dim = random(0, ndim).to(int).value()
-    z = torch.cumprod(x, dim)
+    z = torch.cumprod(y, dim)
```
Comment (MARD1NO) [inline python/oneflow/test/modules/test_global_cumprod.py:30]: 之前yinggang写的一个小bug，在这里fix

File: python/oneflow/test/modules/test_global_groupnorm.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+import oneflow as flow
+import oneflow.unittest
+
+from oneflow.test_utils.automated_test_util import *
+
+
+@autotest(n=1, check_graph=False)
+def _test_global_group_norm(test_case, placement, input_sbp):
+    if placement.type == "cpu":
```
Comment (MARD1NO) [inline python/oneflow/test/modules/test_global_groupnorm.py:31]: cpu还是采取拼的，暂时跳过对应的global test

File: python/oneflow/nn/modules/normalization.py [hunk 5]
```diff
@@ -111,22 +111,32 @@ def forward(self, input: Tensor) -> Tensor:
...
-        normalized = flow.reshape(
-            normalized, shape=[origin_shape[0], self.num_channels, -1]
-        )
-        if self.weight is not None:
-            normalized = normalized * self.weight.reshape(1, self.num_channels, 1)
-        if self.bias is not None:
-            normalized = normalized + self.bias.reshape(1, self.num_channels, 1)
-        res = flow.reshape(normalized, shape=tuple(input.shape))
-        return res
+        if not input.is_cuda:
```
Comment (MARD1NO) [inline python/oneflow/nn/modules/normalization.py:114]: CPU采取的还是拼的方式

File: python/oneflow/test/modules/test_global_groupnorm.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    y = m(x)
+    return y
+
+
+@unittest.skipIf(os.getenv("ONEFLOW_TEST_CPU_ONLY"), "only test cpu cases")
+class TestGroupNormModule(flow.unittest.TestCase):
+    @globaltest
+    def test_global_group_norm_with_random_data(test_case):
+        for placement in all_placement():
+            for sbp in all_sbp(placement, max_dim=1):
```
Comment (MARD1NO) [inline python/oneflow/test/modules/test_global_groupnorm.py:62]: SBP限制为s0，暂时想不到别的sbp了

File: oneflow/user/kernels/group_norm_kernel.cu [hunk 7]
```diff
@@ -0,0 +1,614 @@
...
+  return cudaSuccess;
+}
+
+template<typename T>
+struct SumOp {
+  __device__ __forceinline__ T operator()(const T& a, const T& b) const { return a + b; }
+};
+
+template<typename T, typename ComputeType>
+__global__ void GroupNormParamGradKernel(const T* dy, const T* x, const T* mean, const T* inv_var,
```
Comment (MARD1NO) [inline oneflow/user/kernels/group_norm_kernel.cu:528]:
```python
ds = reduce_sum(dy * x, (2, 3)) # N C
db = reduce_sum(dy, (2, 3)) # N C

dbeta = reduce_sum(db, (0)) # C
dgamma = (ds[nc] - db[nc] * mean[ng]) * inv_var[ng]

```

这里的思路是，一个block处理一个channel。而一个block的线程，处理N, H, W维度上元素部分累加和，然后再做blockreduce，进而得到 reduce_sum((0, 2, 3))的结果



block每个线程

File: oneflow/user/kernels/group_norm_kernel.cu [hunk 8]
```diff
@@ -0,0 +1,699 @@
...
+  using LoadType = PackType<T, PackSize>;
+  for (int32_t channel = blockIdx.x; channel < channel_size; channel += gridDim.x) {
+    ComputeType dgamma_sum = 0.0;
+    ComputeType dbeta_sum = 0.0;
+    const int32_t group_num = channel_size / group_size;
+    for (int32_t batch = 0; batch < batch_size; batch++) {
+      const int32_t batch_channel_id = batch * channel_size + channel;
+      ComputeType ds_sum = 0.0;
+      ComputeType db_sum = 0.0;
+      for (int32_t spatial = threadIdx.x * PackSize; spatial < spatial_size;
```
Comment (MARD1NO) [inline oneflow/user/kernels/group_norm_kernel.cu:584]: spatial 维度是连续的，这里做pack向量化

File: oneflow/user/kernels/group_norm_kernel.cu (hunk 7 above)
Comment (MARD1NO) [reply oneflow/user/kernels/group_norm_kernel.cu:620]: 新版的逻辑是采取2stage reduce。

启动 N * C 个 Block 来做 reduce，处理H W维度。并且把mean 和 var也给参与到累加过程。
后面再启动N*C个线程，在N方向做reduce。

```
reduce(dy * (x - mean)) 
= reduce(dy * x - dy * mean) 
= reduce(dy * x) - reduce(dy * mean) 
= reduce(dy * x) - reduce(dy) * mean
```

File: oneflow/user/ops/groupnorm_op.cpp [hunk 9]
```diff
@@ -0,0 +1,172 @@
+/*
```
[the remark this answers is not in the document]
Comment (MARD1NO) [reply oneflow/user/ops/groupnorm_op.cpp]: 已改

File: oneflow/core/functional/impl/nn_functor.cpp [hunk 10]
```diff
@@ -788,6 +788,44 @@ class LayerNormAffineFunctor {
...
```
[the remark this answers is not in the document]
Comment (MARD1NO) [reply oneflow/core/functional/impl/nn_functor.cpp:775]: 已改

File: python/oneflow/test/modules/test_global_groupnorm.py (hunk 4 above)
Comment (MARD1NO) [reply python/oneflow/test/modules/test_global_groupnorm.py:31]: debug了一下午，在HW都等于1的情况下，oneflow拼的和pytorch自己的kernel有1e-2的误差。如果两边都采取拼的，则没有误差。所以暂时还是跳过了

Comment (yuanms2) [reply python/oneflow/test/modules/test_global_groupnorm.py:31]: 这个找人一块review下

**Comment by MARD1NO** [comment]
global测试还是暂时跳过CPU。经过debug，CPU拼接得到的实现在某些数据下和PyTorch的Groupnorm会有误差。而将PyTorch的GroupNorm替换成拼接实现，则可以正常通过测试

File: oneflow/core/cuda/layer_norm.cuh [hunk 11]
```diff
@@ -161,6 +161,7 @@ struct DirectLoad {
 #pragma unroll
     for (int i = 0; i < N; ++i) { dst[i] = static_cast<DST>(pack.elem[i]); }
   }
+  bool CanPackAs(size_t pack_size) { return true; }
```
Comment (guo-ran) [inline oneflow/core/cuda/layer_norm.cuh:164]: 不需要

File: oneflow/core/cuda/layer_norm.cuh [hunk 12]
```diff
@@ -176,6 +177,7 @@ struct DirectStore {
     for (int i = 0; i < N; ++i) { pack.elem[i] = static_cast<DST>(src[i]); }
     *(reinterpret_cast<PackType<DST, N>*>(dst) + offset) = pack.storage;
   }
+  bool CanPackAs(size_t pack_size) { return true; }
```
Comment (guo-ran) [inline oneflow/core/cuda/layer_norm.cuh:180]: 不需要

File: oneflow/core/cuda/layer_norm.cuh [hunk 13]
```diff
@@ -526,10 +528,10 @@ struct DispatchLayerNormWarpImplPackSize {
   cudaError_t operator()(cudaStream_t stream, LOAD load, STORE store, const int64_t rows,
                          const int64_t cols, const double epsilon, ComputeType* mean,
                          ComputeType* inv_variance) {
-    if (cols % 4 == 0) {
+    if (load.CanPackAs(4) && store.CanPackAs(4)) {
```
Comment (guo-ran) [inline oneflow/core/cuda/layer_norm.cuh:531]: cols % 4 == 0 &&

Comment (guo-ran) [inline oneflow/core/cuda/layer_norm.cuh:531]: 在前面加上这个，只有group_norm的store需要定义CanPackAs 函数，其他的不需要加
``` c++

template<typename T>
class HasCanPackAs {
  typedef char one;
  struct two {
    char x[2];
  };

  template<typename C>
  static one test(decltype(&C::CanPackAs));
  template<typename C>
  static two test(...);

 public:
  enum { value = sizeof(test<T>(0)) == sizeof(char) };
};

template<typename T>
typename std::enable_if<HasCanPackAs<T>::value == true, bool>::type CanPackAs(T t, size_t p) {
  return t.CanPackAs(p);
}

template<typename T>
typename std::enable_if<HasCanPackAs<T>::value == false, bool>::type CanPackAs(T t, size_t p) {
  return true;
}
```

File: oneflow/core/cuda/layer_norm.cuh [hunk 14]
```diff
@@ -526,10 +554,10 @@ struct DispatchLayerNormWarpImplPackSize {
   cudaError_t operator()(cudaStream_t stream, LOAD load, STORE store, const int64_t rows,
                          const int64_t cols, const double epsilon, ComputeType* mean,
                          ComputeType* inv_variance) {
-    if (cols % 4 == 0) {
+    if (cols % 4 == 0 && CanPackAs<STORE>(store, 4)) {
```
Comment (guo-ran) [inline oneflow/core/cuda/layer_norm.cuh:557]: (cols % 4 == 0 && CanPackAs<LOAD>(load, 4) && CanPackAs<STORE>(store, 4))

File: oneflow/user/kernels/group_norm_kernel.cu [hunk 15]
```diff
@@ -0,0 +1,548 @@
...
+        gamma(gamma),
+        beta(beta) {}
+
+  template<int PackSize>
+  __device__ void store(const SRC* src, int64_t row, int64_t col) {
+    cuda::layer_norm::Pack<DST, PackSize> y_pack;
+    const int64_t offset = row * row_size + col;
+    const int64_t packed_offset = offset / PackSize;
+    const int64_t gamma_beta_offset = (offset / spatial_size) % channel_size;
+    DST gamma_val = gamma[gamma_beta_offset];
```
Comment (guo-ran) [inline oneflow/user/kernels/group_norm_kernel.cu:44]: if affine == False时这里会出错吧？

File: python/oneflow/test/modules/test_global_groupnorm.py [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+
+@autotest(n=1, check_graph=False)
+def _test_global_group_norm(test_case, placement, input_sbp):
+    if placement.type == "cpu":
+        return
+    batch_size = 8
+    channel_size = 4
+    num_groups = 2
+    m = torch.nn.GroupNorm(
+        num_groups=num_groups, num_channels=channel_size, affine=True
```
Comment (guo-ran) [inline python/oneflow/test/modules/test_global_groupnorm.py:37]: 最好affine=True 和False时都加个测试吧

File: oneflow/user/kernels/group_norm_kernel.cu [hunk 17]
```diff
@@ -0,0 +1,552 @@
...
+    user_op::Tensor* tmp_buffer = ctx->Tensor4ArgNameAndIndex("tmp_buffer", 0);
+    T* reduce_ds_buf_ptr = reinterpret_cast<T*>(tmp_buffer->mut_dptr<char>());
+    T* reduce_db_buf_ptr =
+        reinterpret_cast<T*>(tmp_buffer->mut_dptr<char>() + batch_size * channel_size * sizeof(T));
+    using ComputeType = typename cuda::layer_norm::DefaultComputeType<T>::type;
+    DispatchGroupNormParamGradKernel<T, ComputeType>(
+        ctx->stream(), dy->dptr<T>(), x->dptr<T>(), mean->dptr<T>(), inv_variance->dptr<T>(),
+        reduce_ds_buf_ptr, reduce_db_buf_ptr, batch_size, group_size, channel_size, spatial_size);
+    int num_blocks;
+    OF_CUDA_CHECK(GetNumBlocks(batch_size * channel_size, &num_blocks));
```
Comment (guo-ran) [inline oneflow/user/kernels/group_norm_kernel.cu:524]: 这里应该是channel_size?

File: oneflow/user/kernels/group_norm_kernel.cu [hunk 18]
```diff
@@ -0,0 +1,552 @@
...
+    const int64_t channel_size = x->shape_view().At(1);
+    const int64_t spatial_size = x->shape_view().elem_cnt() / batch_size / channel_size;
+    const int64_t group_size = num_instances / batch_size;
+    user_op::Tensor* tmp_buffer = ctx->Tensor4ArgNameAndIndex("tmp_buffer", 0);
+    T* reduce_ds_buf_ptr = reinterpret_cast<T*>(tmp_buffer->mut_dptr<char>());
+    T* reduce_db_buf_ptr =
+        reinterpret_cast<T*>(tmp_buffer->mut_dptr<char>() + batch_size * channel_size * sizeof(T));
+    using ComputeType = typename cuda::layer_norm::DefaultComputeType<T>::type;
+    DispatchGroupNormParamGradKernel<T, ComputeType>(
+        ctx->stream(), dy->dptr<T>(), x->dptr<T>(), mean->dptr<T>(), inv_variance->dptr<T>(),
```
Comment (guo-ran) [inline oneflow/user/kernels/group_norm_kernel.cu:521]: mean和inv_variance的dtype不是T吧，应该是ComputeType

File: oneflow/user/kernels/group_norm_kernel.cu (hunk 7 above)
Comment (guo-ran) [reply oneflow/user/kernels/group_norm_kernel.cu:386]: mean和inv_var是ComputeType

**Review by guo-ran — APPROVED** [review]

File: oneflow/user/kernels/group_norm_kernel.cu [hunk 19]
```diff
@@ -0,0 +1,529 @@
...
+        return tmp_buffer_size;                                                           \
+      })                                                                                  \
+      .SetIsMatchedHob((user_op::HobDeviceType() == DeviceType::kCUDA)                    \
+                       && (user_op::HobDataType("dy", 0) == GetDataType<dtype>::value));
+
+REGISTER_GROUP_NORM_PARAM_GRAD_CUDA_KERNEL(half, float)
+REGISTER_GROUP_NORM_PARAM_GRAD_CUDA_KERNEL(float, float)
+REGISTER_GROUP_NORM_PARAM_GRAD_CUDA_KERNEL(double, double)
+#if CUDA_VRSION >= 11000
+REGISTER_GROUP_NORM_PARAM_GRAD_CUDA_KERNEL(nv_bfloat16)
```
Comment (guo-ran) [inline oneflow/user/kernels/group_norm_kernel.cu:526]: 这个怎么只有一个参数

Comment (MARD1NO) [reply oneflow/user/kernels/group_norm_kernel.cu]: 已修改

**Review by liujuncheng — COMMENTED** [review]
还要添加到AMP list里面吧

**Review by liujuncheng — APPROVED** [review]

File: python/oneflow/test/modules/test_global_cumprod.py (hunk 3 above)
Comment (MARD1NO) [reply python/oneflow/test/modules/test_global_cumprod.py:30]: 先暂时revert回去，由yinggang查下具体bug

**Review by guo-ran — APPROVED** [review]
